visidata/loaders/pandas_freqtbl.py

```diff
@@ -0,0 +1,155 @@
+        self.df = df
+        self.mask_bool = mask  # boolean mask
+        self.mask_iloc = np.where(mask.values)[0]  # integer indexes corresponding to mask
+        self.mask_count = mask.sum()
+
+    def __len__(self):
+        return self.mask_count
+
+    def __getitem__(self, k):
+        if isinstance(k, slice):
+            import pandas as pd
+            new_mask = pd.Series(False, index=self.df.index)
+            new_mask.iloc[self.mask_iloc[k]] = True
+            return DataFrameRowSliceAdapter(self.df, new_mask)
+        return self.df.iloc[self.mask_iloc[k]]
+
+    def __iter__(self):
+        return DataFrameRowSliceIter(self.df, self.mask_iloc)
+
+    def __getattr__(self, k):
+        # This is trouble ..
+        return getattr(self.df[self.mask_bool], k)
+
+class DataFrameRowSliceIter:
+    def __init__(self, df, mask_iloc, index=0):
+        self.df = df
+        self.mask_iloc = mask_iloc
+        self.index = index
+
+    def __next__(self):
+        # Accessing row of original dataframe, to ensure
+        # that no copies are made and id() of selected rows
+        # will match original dataframe's rows
+        if self.index >= self.mask_iloc.shape[0]:
+            raise StopIteration()
+        row = self.df.iloc[self.mask_iloc[self.index]]
+        self.index += 1
+        return row
+
+class PandasSheetFreqTable(SheetPivot):
+    'Generate frequency-table sheet on currently selected column.'
+    rowtype = 'bins'  # rowdef FreqRow(keys, sourcerows)
+
+    def __init__(self, sheet, *groupByCols):
+        fqcolname = '%s_%s_freq' % (sheet.name, '-'.join(col.name for col in groupByCols))
+        super().__init__(fqcolname, groupByCols, [], source=sheet)
+        self.largest = 1
+
+    def selectRow(self, row):
+        # raise ValueError(str(type(self.source)) + " " + str(type(row.sourcerows)))
+        self.source.select(row.sourcerows)     # select all entries in the bin on the source sheet
+        return super().selectRow(row)  # then select the bin itself on this sheet
+
+    def unselectRow(self, row):
+        self.source.unselect(row.sourcerows)
+        return super().unselectRow(row)
+
+    def updateLargest(self, grouprow):
+        self.largest = max(self.largest, len(grouprow.sourcerows))
+
+    # @asyncthread
+    def reload(self):
+        'Generate frequency table then reverse-sort by length.'
+        # Conditions:
+        # (1) this only works if len(groupByCols) == 1 since it relies on value_counts()
+        # (2) this assumes some amount of non-degeneracy (non-empty dataframe with non-null values)
+        super().initCols(use_range=False)
+        # columns now the same as the original table
+        # raise ValueError(str([c.name for c in self.columns]))
+
+        df = self.source.rows
+
+        if len(self.groupByCols) > 1:
+            raise NotImplementedError("Frequency Table for > 1 columns currently not supported for PandasSheet")
```

Comment:
I use `error()` or `fail()` for these (both of which raise an ExpectedException).  error() is red; fail() is yellow and less severe.

Reply:
👍 

I think it should be possible to do multiple `groupByCols` here by either using https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html and `stack()`, or by dropping down to `PivotTable`'s implementation. I'm not sure if there are gotchas to using `pd.crosstab().stack()` but I'll give it a shot.